design-proposals/backend-storage-block-migration.md

```diff
@@ -0,0 +1,67 @@
+- RWO block: not supported yet, but PRs exist to implement it in a non-migratable way 
+- RWX block: same as above, with migratability functional but considered unsafe, because it leads to 2 pods (source and target virt-launchers) mounting the same ext3 filesystem concurrently.
+
+This is the best we can do with the current implementation, which involves mounting the same partition on migration source and target.
+This proposal is about switching to a different model which would instead create a new backend-storage PVC on migration,
+allowing all 4 cases above to have proper migration support.
+
+## Motivation/Goals
+- Users want to be able to use any storage class for backend-storage
+- Users want all VMs with persistent features to be potentially migratable
+
+## Definition of Users
+Any VM owner that wants to use persistent features, such as TPM or EFI
+
+## User Stories
+As a user, I want to be able to seamlessly enable persistent features while keeping the migratability of my VMs.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+- When creating a new VM with persistent features, we use either (first match wins):
+  - The storage class specified in the KubeVirt Custom Resource
+  - The storage class marked as the default for virtualization purposes
+  - The storage class marked as the default for the cluster
+- All new persistent storage PVCs will be created with the name `persistent-state-for-<vm_name>-<random_string>` and the new annotation `persistent-state-for-<vm_name>=true`
+- When starting an existing VM with persistent features, we will look for any PVC with the annotation, or any PVC with the legacy name, for backwards compatibility
+- For volume/access modes:
+  - When multiple volume modes are listed in the StorageProfile of the StorageClass, we will default to filesystem (to avoid partitioning complexity)
+  - When multiple access modes are listed for the selected volume mode, we will arbitrarily default to RWO.
+  - When CDI is not present, or the StorageClass has no StorageProfile, or the StorageProfile has no modes, we'll default to RWO FS.
+- On migration:
+  - No matter what volume/access mode the PVC uses, we will create a new empty PVC for the target with the name `persistent-state-for-<vm_name>-<random_string>` and no annotation
+  - In the migration object status, we store the names of both the source and target PVCs (see API section below)
+  - If a migration succeeds, we set the annotation `persistent-state-for-<vm_name>=true` on the new (target) PVC and delete the old (source) PVC, using the source PVC name from the migration object
+  - If a migration fails, we delete the target backend-storage PVC that was just created, using the target PVC name from the migration object
+  - In the unlifely event that a VM shuts down towards the very end of a migration, the migration finalizer will decide which PVC to keep and which one to get rid of
```

Comment:
how could the finalizer make this decision? I don't think it has enough information.

Reply:
The finalizer itself doesn't solve this issue. It only makes sure that the migration object won't get deleted before the correct volume is annotated by the controller.

However, there is still a race that this proposal will need to address. 

There is a short period when the migration is finalized on the host, but before the controller has annotated the correct volume. The VM could shutdown and start again, picking up an incorrect volume.

I think that this issue can be addressed by simply re-queueing the newly created VMI object if it has an unfinalized migration. 
This is an issue on its own that a new VM can start while it has an unfinished migration.

Reply:
Yes, I agree. @jean-edouard could you please highlight this race and the suggested solution?